File: packages/kits/NeuronKit/rust/src/context_synthesizer.rs
//! ContextSynthesizer Rust version. Pure synthesis over a recall page —
//! reads only, produces a `ContextDocument` matching the Swift
//! `ContextSynthesisEngine` byte-for-byte against shared vectors.
//!
//! Per NEURONKIT_SPEC § 4.2 the synthesizer is reads-only and writes
//! nothing to the substrate (C-9). The Rust version mirrors this by
//! exposing only a pure function over a recall page; no estate
//! handle, no IO, no clocks.

use crate::hybrid_recall::{DrawerRow, RecallPage};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// Ephemeral context document. Mirror of Swift's `ContextDocument`.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct ContextDocument {
    pub summary: String,
    pub patterns: Vec<String>,
    pub success_rate: f32,
    pub average_reward: f32,
    pub recommendations: Vec<String>,
    pub key_insights: Vec<String>,
}

/// Optional row-level metadata the Rust engine consumes alongside
/// `DrawerRow`. The Swift version reads these straight off
/// `LocusKit.Drawer`; the Rust engine accepts them as a parallel
/// vector so the version can be exercised without the full LocusKit
/// Rust type. Empty (or shorter than `rows`) means "treat as default
/// (active state, default wing/room)".
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct DrawerRowMeta {
    /// Parent node id — the node-tree anchor the summary names. The Swift engine
    /// reads `LocusKit.Drawer.parentNodeId`; the Rust version accepts it as
    /// parallel metadata. Drives `make_summary`'s "dominant node {id}".
    pub parent_node_id: String,
    /// Wing/room remain as vestigial metadata (the summary no longer reads them
    /// after the node-tree migration; kept for parity with the Swift fixture,
    /// which still carries them).
    pub wing: String,
    pub room: String,
    /// Whether the drawer is "currently believed" per the LocusKit
    /// adjective-state cluster. The Swift engine reads
    /// `drawer.isCurrentlyBelieved`; the Rust version accepts the
    /// pre-computed boolean.
    pub is_currently_believed: bool,
}

impl Default for DrawerRowMeta {
    fn default() -> Self {
        Self {
            parent_node_id: "(no node)".to_string(),
            wing: "(no wing)".to_string(),
            room: "(no room)".to_string(),
            is_currently_believed: true,
        }
    }
}

/// Synthesize a `ContextDocument` from a recall page and the
/// per-row metadata vector. The metadata vector may be empty, in
/// which case every row is treated with `DrawerRowMeta::default()`.
///
/// `max_key_insights` bounds the excerpted rows. The historical digest
/// bound is 3. A cue-grounded caller passes its post-rank cap so every
/// ranked survivor is VISIBLE in the document — trial 3 measured 30/35
/// misses with the answer drawer ranked into the capped set but invisible
/// behind the 3-row excerpt. Twin of Swift
/// `ContextSynthesisEngine.synthesize(page:maxKeyInsights:)`.
pub fn synthesize(
    page: &RecallPage,
    meta: &[DrawerRowMeta],
    max_key_insights: usize,
) -> ContextDocument {
    let rows = &page.rows;
    if rows.is_empty() {
        return ContextDocument {
            summary: String::new(),
            patterns: Vec::new(),
            success_rate: 0.0,
            average_reward: 0.0,
            recommendations: Vec::new(),
            key_insights: Vec::new(),
        };
    }

    let summary = make_summary(rows, meta);
    let patterns = top_patterns(rows, 5);
    let success_rate = currently_believed_rate(rows, meta);
    let average_reward: f32 = 0.0; // No reward field on DrawerRow at v0.1 — see spec note.
    let recommendations = make_recommendations(&patterns);
    let key_insights = make_key_insights(rows, max_key_insights.max(1));

    ContextDocument {
        summary,
        patterns,
        success_rate,
        average_reward,
        recommendations,
        key_insights,
    }
}

fn meta_or_default(meta: &[DrawerRowMeta], idx: usize) -> DrawerRowMeta {
    meta.get(idx).cloned().unwrap_or_default()
}

/// One-line summary naming row count and the dominant parent node. Stable
/// across runs. Mirrors Swift `ContextSynthesisEngine.makeSummary`, which reads
/// `Drawer.parentNodeId` (node-tree anchor) — not wing/room.
pub fn make_summary(rows: &[DrawerRow], meta: &[DrawerRowMeta]) -> String {
    let count = rows.len();
    let nodes: Vec<String> = (0..rows.len())
        .map(|i| meta_or_default(meta, i).parent_node_id)
        .collect();
    let top_node = most_frequent(&nodes).unwrap_or_else(|| "(no node)".to_string());
    format!("{} drawers; dominant node {}.", count, top_node)
}

/// Standard English stopwords excluded from pattern extraction. High-frequency
/// function words and date literals provide no semantic signal and would
/// otherwise dominate the pattern list when present in a corpus.
///
/// Mirrors the Swift `ContextSynthesisEngine.stopwords` set byte-for-byte so
/// conformance test vectors produce identical output on both ports.
const STOPWORDS: &[&str] = &[
    "this", "that", "with", "from", "they", "them", "their", "there",
    "were", "have", "been", "will", "would", "could", "should", "about",
    "when", "then", "than", "also", "into", "your", "more", "some",
    "what", "which", "these", "those", "just", "like", "over", "such",
    "only", "very", "even", "most", "both", "each", "here", "after",
    "well", "back", "much", "many", "make", "time", "know", "take",
    "long", "made", "come", "want", "used", "same", "need",
];

/// True when `token` is a bare 4-digit year (1000–2999) or a pure numeric
/// string — neither carries semantic meaning as a pattern. Mirrors Swift
/// `ContextSynthesisEngine.isBareYearOrNumeric(_:)`.
///
/// Uses `char::is_numeric()` (Unicode numeric category: digits, superscripts,
/// fractions, circled numerals, ...) to match Swift's `Character.isNumber`,
/// which is the Unicode general category "Number". Using `is_ascii_digit()`
/// (0-9 only) would silently pass through Unicode numeric tokens that Swift
/// filters, causing cross-port divergence on non-ASCII input. (NK-3 planned hardening)
fn is_bare_year_or_numeric(token: &str) -> bool {
    if !token.chars().all(|c| c.is_numeric()) {
        return false;
    }
    // All-numeric strings of any length are pure numeric — exclude.
    true
}

/// Top-N patterns. Frequency descending, ties broken by first
/// appearance, final tiebreak by token string. Identical ordering
/// to Swift.
///
/// Excludes stopwords and bare numeric strings (including years)
/// so high-frequency function words and date literals do not dominate
/// the pattern list. Mirrors Swift `ContextSynthesisEngine.topPatterns`.
pub fn top_patterns(rows: &[DrawerRow], max_count: usize) -> Vec<String> {
    let mut counts: BTreeMap<String, usize> = BTreeMap::new();
    let mut first_seen: BTreeMap<String, usize> = BTreeMap::new();
    let mut order = 0usize;
    for row in rows {
        for token in tokens(&row.content) {
            if token.chars().count() < 4 {
                continue;
            }
            // Skip stopwords and bare numeric strings (including years).
            if STOPWORDS.contains(&token.as_str()) || is_bare_year_or_numeric(&token) {
                continue;
            }
            *counts.entry(token.clone()).or_insert(0) += 1;
            first_seen.entry(token.clone()).or_insert_with(|| {
                let v = order;
                order += 1;
                v
            });
        }
    }
    let mut keys: Vec<String> = counts.keys().cloned().collect();
    keys.sort_by(|lhs, rhs| {
        let cl = counts.get(lhs).copied().unwrap_or(0);
        let cr = counts.get(rhs).copied().unwrap_or(0);
        if cl != cr {
            return cr.cmp(&cl);
        }
        let fl = first_seen.get(lhs).copied().unwrap_or(usize::MAX);
        let fr = first_seen.get(rhs).copied().unwrap_or(usize::MAX);
        if fl != fr {
            return fl.cmp(&fr);
        }
        lhs.cmp(rhs)
    });
    keys.into_iter().take(max_count).collect()
}

/// Fraction of rows whose meta marks them "currently believed".
/// Treats missing meta as default (true).
pub fn currently_believed_rate(rows: &[DrawerRow], meta: &[DrawerRowMeta]) -> f32 {
    if rows.is_empty() {
        return 0.0;
    }
    let count = (0..rows.len())
        .filter(|&i| meta_or_default(meta, i).is_currently_believed)
        .count();
    count as f32 / rows.len() as f32
}

/// Recommendations from the top patterns; neutral fallback if none.
pub fn make_recommendations(patterns: &[String]) -> Vec<String> {
    if patterns.is_empty() {
        return vec![
            "No dominant pattern detected; consider broadening the recall frame.".to_string(),
        ];
    }
    patterns
        .iter()
        .map(|p| format!("Explore further evidence about '{}'.", p))
        .collect()
}

/// First-line excerpts from up to `max_count` rows.
pub fn make_key_insights(rows: &[DrawerRow], max_count: usize) -> Vec<String> {
    rows.iter()
        .take(max_count)
        .map(|row| match row.content.find('\n') {
            Some(idx) => row.content[..idx].to_string(),
            None => row.content.clone(),
        })
        .collect()
}

/// Most frequent element with ties broken by first appearance.
fn most_frequent(values: &[String]) -> Option<String> {
    if values.is_empty() {
        return None;
    }
    let mut counts: BTreeMap<String, usize> = BTreeMap::new();
    let mut first_seen: BTreeMap<String, usize> = BTreeMap::new();
    let mut order = 0usize;
    for v in values {
        *counts.entry(v.clone()).or_insert(0) += 1;
        first_seen.entry(v.clone()).or_insert_with(|| {
            let cur = order;
            order += 1;
            cur
        });
    }
    let mut best: Option<(String, usize, usize)> = None;
    for (k, c) in &counts {
        let fs = first_seen.get(k).copied().unwrap_or(usize::MAX);
        match &best {
            None => best = Some((k.clone(), *c, fs)),
            Some((_, bc, bfs)) => {
                if c > bc || (c == bc && fs < *bfs) {
                    best = Some((k.clone(), *c, fs));
                }
            }
        }
    }
    best.map(|(k, _, _)| k)
}

/// Lowercase alphanumeric tokens. Locale-free; identical to Swift's
/// `tokens` over ASCII conformance vectors.
pub fn tokens(s: &str) -> Vec<String> {
    let lower = s.to_lowercase();
    lower
        .split(|c: char| !(c.is_alphabetic() || c.is_numeric()))
        .filter(|t| !t.is_empty())
        .map(String::from)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn row(content: &str) -> DrawerRow {
        DrawerRow {
            id: format!("id-{}", content.len()),
            content: content.to_string(),
        }
    }

    fn meta(wing: &str, room: &str, believed: bool) -> DrawerRowMeta {
        DrawerRowMeta {
            parent_node_id: "(no node)".to_string(),
            wing: wing.to_string(),
            room: room.to_string(),
            is_currently_believed: believed,
        }
    }

    fn meta_node(node: &str) -> DrawerRowMeta {
        DrawerRowMeta {
            parent_node_id: node.to_string(),
            ..DrawerRowMeta::default()
        }
    }

    #[test]
    fn empty_page_produces_empty_document() {
        let page = RecallPage {
            rows: vec![],
            page_index: 0,
            is_last: true,
        };
        let doc = synthesize(&page, &[], 3);
        assert_eq!(doc.summary, "");
        assert!(doc.patterns.is_empty());
        assert_eq!(doc.success_rate, 0.0);
        assert_eq!(doc.average_reward, 0.0);
        assert!(doc.recommendations.is_empty());
        assert!(doc.key_insights.is_empty());
    }

    #[test]
    fn summary_names_count_and_dominant_node() {
        let rows = vec![row("a"), row("b"), row("c")];
        let m = vec![
            meta_node("node-x"),
            meta_node("node-x"),
            meta_node("node-y"),
        ];
        let page = RecallPage {
            rows,
            page_index: 0,
            is_last: true,
        };
        let doc = synthesize(&page, &m, 3);
        assert_eq!(doc.summary, "3 drawers; dominant node node-x.");
    }

    #[test]
    fn patterns_rank_by_frequency_then_first_seen() {
        let rows = vec![
            row("carbon organic compounds"),
            row("organic chemistry carbon"),
            row("physics waves photons"),
        ];
        let page = RecallPage {
            rows,
            page_index: 0,
            is_last: true,
        };
        let doc = synthesize(&page, &[], 3);
        assert_eq!(
            doc.patterns,
            vec!["carbon", "organic", "compounds", "chemistry", "physics"]
        );
    }

    #[test]
    fn recommendations_match_pattern_count() {
        let rows = vec![row("alpha beta gamma delta")];
        let page = RecallPage {
            rows,
            page_index: 0,
            is_last: true,
        };
        let doc = synthesize(&page, &[], 3);
        assert_eq!(doc.recommendations.len(), doc.patterns.len());
    }

    #[test]
    fn no_pattern_produces_neutral_recommendation() {
        let rows = vec![row("a bb ccc")];
        let page = RecallPage {
            rows,
            page_index: 0,
            is_last: true,
        };
        let doc = synthesize(&page, &[], 3);
        assert!(doc.patterns.is_empty());
        assert_eq!(doc.recommendations.len(), 1);
        assert!(doc.recommendations[0].contains("broadening the recall frame"));
    }

    #[test]
    fn key_insights_take_first_line_up_to_three_rows() {
        let rows = vec![
            row("line one\nbody body"),
            row("single line"),
            row("three\nthree body"),
            row("four — should not appear"),
        ];
        let page = RecallPage {
            rows,
            page_index: 0,
            is_last: true,
        };
        let doc = synthesize(&page, &[], 3);
        assert_eq!(doc.key_insights, vec!["line one", "single line", "three"]);
    }

    #[test]
    fn currently_believed_rate_counts_fraction() {
        let rows = vec![row("a"), row("b"), row("c")];
        let m = vec![
            meta("w", "r", true),
            meta("w", "r", true),
            meta("w", "r", false),
        ];
        let page = RecallPage {
            rows,
            page_index: 0,
            is_last: true,
        };
        let doc = synthesize(&page, &m, 3);
        assert!((doc.success_rate - (2.0 / 3.0)).abs() < 1e-6);
    }
}

File: packages/kits/NeuronKit/Sources/NeuronKit/ContextSynthesizer.swift
// ContextSynthesizer.swift
//
// Reads-only synthesis of a `RecallStream.Page` into a
// `ContextDocument` suitable for foundation-model consumption. Per
// NEURONKIT_SPEC § 4.2 and invariant C-9, the synthesizer never
// writes to the substrate and calls no estate verbs. Pure, ephemeral.
//
// The `estate:` parameter is preserved from the spec signature for
// shape parity with § 4.2 and to keep future evolution open (the
// boundary can grow read-only queries without forcing every caller
// to add an argument). Under C-9 the parameter is reserved and
// untouched — verb dispatch through `glk.recall` or any other verb
// is prohibited here.

import Foundation
import GeniusLocusKit
import LocusKit

/// Ephemeral context document produced by `ContextSynthesizer`. Per
/// spec § 4.2 — handed to a foundation model, never written to the
/// substrate, never persisted. Equatable for test conformance and
/// snapshot parity against the Rust version.
public struct ContextDocument: Sendable, Equatable, Codable {

    /// Short prose summary of the recalled set (one or two sentences).
    /// Deterministic — derived from the drawer content shingles, not
    /// from a model.
    public let summary: String

    /// Salient repeating patterns observed across the recalled set.
    /// Stable order: descending by occurrence, ties broken by first
    /// appearance.
    public let patterns: [String]

    /// Proportion of recalled drawers whose adjective-bitmap `state`
    /// is the "currently believed" cluster per LocusKit spec § 6.1.
    /// In [0, 1]; reported as Float so the round-trip is identical
    /// to the Rust f32 version.
    public let successRate: Float

    /// Mean reward across the recalled set. Drawer.reward is not a
    /// substrate field today, so this value is 0.0 for v0.1 callers.
    /// Carried for shape parity with the spec; the field stays in
    /// the public surface so the day a reward attribute lands the
    /// caller does not change.
    public let averageReward: Float

    /// Recommendations distilled from the page. Deterministic strings
    /// built from the dominant patterns; never invented by a model.
    public let recommendations: [String]

    /// Key insights — first-line excerpts from the highest-relevance
    /// drawers in the page, in stream order.
    public let keyInsights: [String]

    public init(
        summary: String,
        patterns: [String],
        successRate: Float,
        averageReward: Float,
        recommendations: [String],
        keyInsights: [String]
    ) {
        self.summary = summary
        self.patterns = patterns
        self.successRate = successRate
        self.averageReward = averageReward
        self.recommendations = recommendations
        self.keyInsights = keyInsights
    }
}

/// Namespace for the synthesis function. Declared as an enum (no
/// cases) so it cannot be instantiated and so test code references
/// the function with `ContextSynthesizer.synthesize(...)`, matching
/// the spec's prose.
public enum ContextSynthesizer {

    /// Synthesize a `ContextDocument` from one `RecallStream.Page`.
    ///
    /// **C-9 invariant:** This function performs no estate write,
    /// invokes no estate verb, and accesses no substrate state
    /// beyond what is already materialised in `page.rows`. The
    /// `estate:` parameter is reserved and untouched — present to
    /// match the spec signature and to allow future read-only verb
    /// composition without a source break for callers. The compiler
    /// will warn if it is unused; the underscore-discard below
    /// records the intentional non-use so future maintainers see
    /// the invariant in code.
    public static func synthesize(
        from page: RecallStream.Page,
        estate: EstateHandle
    ) async throws -> ContextDocument {
        _ = estate // C-9: reserved, never consulted. See header doc.
        return ContextSynthesisEngine.synthesize(page: page)
    }
}

/// Pure synthesis engine. Module-internal so the Swift conformance
/// tests and the Rust version exercise the same deterministic math
/// against shared vectors. No `async`, no estate handle — just the
/// drawer rows.
internal enum ContextSynthesisEngine {

    static func synthesize(page: RecallStream.Page) -> ContextDocument {
        let rows = page.rows
        if rows.isEmpty {
            return ContextDocument(
                summary: "",
                patterns: [],
                successRate: 0,
                averageReward: 0,
                recommendations: [],
                keyInsights: []
            )
        }

        let summary = makeSummary(rows: rows)
        let patterns = topPatterns(rows: rows, maxCount: 5)
        let successRate = currentlyBelievedRate(rows: rows)
        let averageReward: Float = 0 // No reward field on Drawer at v0.1 — see spec note.
        let recommendations = makeRecommendations(patterns: patterns)
        let keyInsights = makeKeyInsights(rows: rows, maxCount: 3)

        return ContextDocument(
            summary: summary,
            patterns: patterns,
            successRate: successRate,
            averageReward: averageReward,
            recommendations: recommendations,
            keyInsights: keyInsights
        )
    }

    /// Build a one-line summary that names the page row count, the
    /// most frequent parent node, and the dominant content-kind. Stable
    /// across runs — no clocks, no randomness.
    static func makeSummary(rows: [Drawer]) -> String {
        let count = rows.count
        let topNode = mostFrequent(rows.map { $0.parentNodeId }) ?? "(no node)"
        return "\(count) drawers; dominant node \(topNode)."
    }

    /// Standard English stopwords excluded from pattern extraction. These high-
    /// frequency function words and bare 4-digit years add no semantic signal
    /// and would dominate the pattern list when present in a corpus.
    ///
    /// Mirrors the Rust `STOPWORDS` constant in `context_synthesizer.rs`
    /// byte-for-byte so conformance test vectors produce identical output.
    static let stopwords: Set<String> = [
        "this", "that", "with", "from", "they", "them", "their", "there",
        "were", "have", "been", "will", "would", "could", "should", "about",
        "when", "then", "than", "also", "into", "your", "more", "some",
        "what", "which", "these", "those", "just", "like", "over", "such",
        "only", "very", "even", "most", "both", "each", "here", "after",
        "well", "back", "much", "many", "make", "time", "know", "take",
        "long", "made", "come", "want", "used", "same", "need",
    ]

    /// True when `token` is a bare 4-digit year (1000–2999) or a pure numeric
    /// string — neither carries semantic meaning as a pattern.
    private static func isBareYearOrNumeric(_ token: String) -> Bool {
        guard token.allSatisfy({ $0.isNumber }) else { return false }
        if token.count == 4,
           let year = Int(token), year >= 1000, year <= 2999 { return true }
        // Pure numeric strings (any length) are also excluded.
        return true
    }

    /// Tokenise each drawer's content into 4-or-longer lowercase
    /// alphanumeric words, then return the most frequent ones across
    /// the page in descending order. Stable: ties break by first
    /// appearance.
    ///
    /// Excludes stopwords and bare numeric strings (including years)
    /// so high-frequency function words and date literals do not dominate
    /// the pattern list. Mirrors the Rust `top_patterns` filter.
    static func topPatterns(rows: [Drawer], maxCount: Int) -> [String] {
        var counts: [String: Int] = [:]
        var firstSeen: [String: Int] = [:]
        var insertionOrder = 0
        for row in rows {
            for token in tokens(row.content) where token.count >= 4
                && !stopwords.contains(token)
                && !isBareYearOrNumeric(token) {
                counts[token, default: 0] += 1
                if firstSeen[token] == nil {
                    firstSeen[token] = insertionOrder
                    insertionOrder += 1
                }
            }
        }
        // Order: count desc, then first-seen asc (stable). Use the
        // token string as a final tiebreaker so output is identical
        // across language versions even when counts and first-seen ties
        // align.
        let ordered = counts.keys.sorted { lhs, rhs in
            let cl = counts[lhs] ?? 0
            let cr = counts[rhs] ?? 0
            if cl != cr { return cl > cr }
            let fl = firstSeen[lhs] ?? Int.max
            let fr = firstSeen[rhs] ?? Int.max
            if fl != fr { return fl < fr }
            return lhs < rhs
        }
        return Array(ordered.prefix(maxCount))
    }

    /// Fraction of rows whose adjective-bitmap `state` is in the
    /// "currently believed" cluster. Computed via LocusKit's
    /// extension predicate `isCurrentlyBelieved` so the math stays
    /// aligned with the substrate definition.
    static func currentlyBelievedRate(rows: [Drawer]) -> Float {
        guard !rows.isEmpty else { return 0 }
        let count = rows.reduce(into: 0) { acc, row in
            if row.isCurrentlyBelieved { acc += 1 }
        }
        return Float(count) / Float(rows.count)
    }

    /// Build deterministic recommendations from the top patterns. For
    /// each pattern, emit one recommendation; if no patterns surfaced,
    /// return a single neutral string so the caller never sees an
    /// empty list when the page itself was non-empty.
    static func makeRecommendations(patterns: [String]) -> [String] {
        guard !patterns.isEmpty else {
            return ["No dominant pattern detected; consider broadening the recall frame."]
        }
        return patterns.map { "Explore further evidence about '\($0)'." }
    }

    /// First-line excerpts from up to `maxCount` rows, in stream
    /// order. The "first line" is the substring up to the first
    /// newline, or the full content if there is no newline.
    static func makeKeyInsights(rows: [Drawer], maxCount: Int) -> [String] {
        rows.prefix(maxCount).map { row in
            if let nl = row.content.firstIndex(of: "\n") {
                return String(row.content[..<nl])
            }
            return row.content
        }
    }

    // MARK: - private helpers

    /// Most frequent element in a sequence, ties broken by first
    /// appearance. Nil for an empty sequence.
    static func mostFrequent<T: Hashable>(_ values: [T]) -> T? {
        guard !values.isEmpty else { return nil }
        var counts: [T: Int] = [:]
        var firstSeen: [T: Int] = [:]
        var order = 0
        for v in values {
            counts[v, default: 0] += 1
            if firstSeen[v] == nil {
                firstSeen[v] = order
                order += 1
            }
        }
        return counts.max { lhs, rhs in
            if lhs.value != rhs.value { return lhs.value < rhs.value }
            return (firstSeen[lhs.key] ?? Int.max) > (firstSeen[rhs.key] ?? Int.max)
        }?.key
    }

    /// Lowercase alphanumeric tokens. Splits on any non-letter,
    /// non-digit character so behaviour is locale-free and identical
    /// across versions for the ASCII conformance vectors used in tests.
    static func tokens(_ s: String) -> [String] {
        s.lowercased()
            .split(whereSeparator: { !($0.isLetter || $0.isNumber) })
            .map(String.init)
    }
}
